### PR TITLE
[TE] Detection and alert config formatters

### PR DESCRIPTION
**Pull request by jihaozh**
This PR adds the Detection config and the detection alert config formatter. These classes includes the business logic for UI presentation. 

- Extract nested metric urns from the nested properties. This enables the frontend to show the metrics for entity alerts. 
- Show different default window size for different detection granularities.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.Map;
+import org.apache.pinot.thirdeye.datalayer.dto.AbstractDTO;
+
+
+/**
+ * Class for building UI formatters for DTOs. Takes in a DTO and returns
+ * a map that contains human-readable data for display on the UI.
+ */
+public interface DTOFormatter<T extends AbstractDTO> {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java]: If it is a general purpose formatter, would be better to put into a separate package.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    } catch (Exception e) {
+      LOG.warn("Exception thrown when getting granularities for detection config {}, use default presenting window size", config.getId(), e);
+      return TimeUnit.DAYS.toMillis(30);
+    }
+  }
+
+  private List<TimeGranularity> getMonitoringGranularities(DetectionConfigDTO config) {
+    List<TimeGranularity> monitoringGranularities = new ArrayList<>();
+    for (Object specs : config.getComponentSpecs().values()) {
+      if (((Map<String, Object>) specs).containsKey(PROP_MONITORING_GRANULARITY)) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: For algorithm the granularity is set by bucketPeriod.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        monitoringGranularities.add(
+            TimeGranularity.fromString(MapUtils.getString((Map<String, Object>) specs, (PROP_MONITORING_GRANULARITY))));
+      }
+    }
+    return monitoringGranularities;
+  }
+
+  private TimeGranularity getGranularityForMetricUrn(String metricUrn) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: It should be detection granularity not metric  granularity.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    MetricEntity me = MetricEntity.fromURN(metricUrn);
+    return this.datasetDAO.findByDataset(this.metricDAO.findById(me.getId()).getDataset()).bucketTimeGranularity();
+  }
+
+  private long getDefaultWindowSizeForGranularity(TimeGranularity granularity) {
+    TimeUnit unit = granularity.getUnit();
+    if (unit == TimeUnit.MINUTES || unit == TimeUnit.HOURS) {
+      return TimeUnit.HOURS.toMillis(48);
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: Convert the numbers to final constants?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java (hunk 2 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: I think it's a bit hacky to put the logic for a specific rule here. For now, I can add it to make it work. Maybe we need to revisit once we have the standard parameters parameter specs.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java (hunk 3 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: Yes. This is only triggered only if the detection granularity is not overridden in the spec. By default, the detection granularity is the metric granularity.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java (hunk 1 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java]: Good idea. Will move to thirdeye.formatter package.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  }
+
+  private List<TimeGranularity> getMonitoringGranularities(DetectionConfigDTO config) {
+    List<TimeGranularity> monitoringGranularities = new ArrayList<>();
+    for (Object specs : config.getComponentSpecs().values()) {
+      if (((Map<String, Object>) specs).containsKey(PROP_MONITORING_GRANULARITY)) {
+        monitoringGranularities.add(
+            TimeGranularity.fromString(MapUtils.getString((Map<String, Object>) specs, (PROP_MONITORING_GRANULARITY))));
+      }
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: What's the behavior if monitoring granularity is not set on all components?

Maybe if monitoring granularity is not set, then add default granularity here?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    List<Map<String, Object>> nestedProperties = ConfigUtils.getList(properties.get(PROP_NESTED_PROPERTIES_KEY));
+    // extract the metric urns recursively from the nested properties
+    for (Map<String, Object> nestedProperty : nestedProperties) {
+      metricUrns.addAll(extractMetricUrnsFromProperties(nestedProperty));
+    }
+    return metricUrns;
+  }
+
+  private long getAlertDetailsDefaultWindowSize(DetectionConfigDTO config, List<String> metricUrns) {
```
[the remark this answers is not in the document]
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: added

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java [hunk 7]
```diff
@@ -108,8 +110,9 @@
   private final DetectionConfigManager configDAO;
   private final EvaluationManager evaluationDAO;
   private final TaskManager taskDAO;
-
   private final DetectionAlertConfigManager detectionAlertConfigDAO;
+  private final DetectionConfigFormatter detectionConfigFormatter;
+  private final DetectionAlertConfigFormatter detectionAlertConfigFormatter;
```
[the remark this answers is not in the document]
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java]: sure. updated

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java (hunk 4 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: updated

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java (hunk 5 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: If monitoring granularity is not set, we will get the default detection granularity from the metric's native granularity.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java (hunk 1 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DTOFormatter.java]: Agree. I am writing a parameter spec.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java (hunk 2 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/formatter/DetectionConfigFormatter.java]: Agree. I am writing a parameter spec.

**Pull request merged**